core/primitives/src/stateless_validation.rs

```diff
@@ -17,10 +18,36 @@ use near_primitives_core::types::{AccountId, Balance, BlockHeight, ShardId};
 /// This is a messy workaround until we know what to do with NEP 483.
 type SignatureDifferentiator = String;
 
-/// Signable
+/// Represents bytes of encoded ChunkStateWitness.
+/// For now encoding is raw borsh serialization, later we plan
+/// adding compression on top of that.
 #[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
-pub struct ChunkStateWitness {
-    pub inner: ChunkStateWitnessInner,
+pub struct EncodedChunkStateWitness(Box<[u8]>);
+
+impl EncodedChunkStateWitness {
+    pub fn encode(witness: &ChunkStateWitness) -> Self {
+        Self(borsh::to_vec(witness).expect("borsh serialization should not fail").into())
+    }
+
+    pub fn decode(&self) -> Result<ChunkStateWitness, std::io::Error> {
+        ChunkStateWitness::try_from_slice(&self.0)
+    }
+
+    pub fn size_bytes(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        &self.0
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
+pub struct SignedChunkStateWitness {
```

Comment:
nit: Could we please name the struct to `SignedEncodedChunkStateWitness`? We can still continue to call the variables signed_witness but having the naming like this really helps establish the hierarchy in my head that signed witness actually contains encoded state witness that later needs decoding to get the actual proper state witness.

Reply:
done: 8a5132956cf3e220d5f823891c54d0ac08b95713